In [44]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [45]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [46]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [47]:
import warnings  # Ignore warnings
warnings.filterwarnings("ignore", category=FutureWarning, module="holoviews")


# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [48]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
scaler = StandardScaler()

# Fit and transform the data
df_market_data_normalized = scaler.fit_transform(df_market_data)

# Create a new DataFrame with the normalized data
df_normalized = pd.DataFrame(df_market_data_normalized, columns=df_market_data.columns, index=df_market_data.index)

df_market_data_normalized[0:5]

array([[ 0.50852937,  0.49319307,  0.77220043,  0.23545963, -0.0674951 ,
        -0.35595348, -0.25163688],
       [ 0.18544589,  0.93444504,  0.55869212, -0.05434093, -0.27348273,
        -0.11575947, -0.19935211],
       [ 0.02177396, -0.70633685, -0.02168042, -0.06103015,  0.00800452,
        -0.55024692, -0.28206051],
       [-0.04076438, -0.81092807,  0.24945797, -0.05038797, -0.37316402,
        -0.45825882, -0.29554614],
       [ 1.19303608,  2.00095907,  1.76061001,  0.54584206, -0.29120287,
        -0.49984776, -0.27031695]])

In [49]:
# Create a DataFrame with the scaled data
df_scaled = pd.DataFrame(df_market_data_normalized, columns= df_market_data.columns)

# Copy the crypto names from the original data
df_scaled['coin_id'] = df_market_data.index

# Set the coinid column as index
df_scaled.set_index('coin_id', inplace=True)

# Display sample data
df_scaled.head()


price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [50]:
# Create a list with the number of k-values from 1 to 11
k_values = list(range(1,12))

In [51]:
# Create an empty list to store the inertia values
inertia_values= []

# Create a for loop to compute the inertia with each possible value of k
for k in k_values:
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
    kmeans_model = KMeans(n_clusters=k, random_state=1, n_init=10)
# 2. Fit the model to the data using `df_market_data_scaled`
    kmeans_model.fit(df_scaled)
# 3. Append the model.inertia_ to the inertia list
    inertia_values.append(kmeans_model.inertia_)

print(inertia_values)

c:\Users\artur\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\artur\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\artur\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\artur\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Window

[287.0, 195.82021818036043, 123.19048183836958, 79.02243535120975, 65.40592346140596, 52.93355892101499, 47.98312409811001, 37.28818726271726, 33.06168486647882, 28.77975243142922, 25.248499378382775]


In [52]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {
    'k_values': k_values,
    'inertia_values': inertia_values
}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow_curve = pd.DataFrame(elbow_data)
df_elbow_curve

k_values  inertia_values
0          1      287.000000
1          2      195.820218
2          3      123.190482
3          4       79.022435
4          5       65.405923
5          6       52.933559
6          7       47.983124
7          8       37.288187
8          9       33.061685
9         10       28.779752
10        11       25.248499

In [53]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
df_elbow_curve.hvplot.line(
    x='k_values',
    y='inertia_values',
    title='Elbow Curve',
    xticks=k_values
)

:Curve   [k_values]   (inertia_values)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 4

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [54]:
# Initialize the K-Means model using the best value for k
kmens_model = KMeans(n_clusters=4 , random_state=1, n_init=10)

In [55]:
# Fit the K-Means model using the scaled data
kmeans_model.fit(df_scaled)

c:\Users\artur\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=11, n_init=10, random_state=1)

In [56]:
# Predict the clusters to group the cryptocurrencies using the scaled data
cluster_labels = kmeans_model.predict(df_scaled)

# Print the resulting array of cluster values.
cluster_labels

array([ 3,  3,  1,  1,  7,  7,  7,  7,  3,  1, 10,  1,  1,  5,  1,  3,  1,
        1,  7,  1, 10,  3,  1,  1,  1,  1,  3, 10,  3,  2,  3,  1,  0,  3,
        8,  6,  4, 10,  3,  1,  9])

In [57]:
# Create a copy of the DataFrame
df_scaled_copy = df_scaled.copy()

In [58]:
# Add a new column to the DataFrame with the predicted clusters
df_scaled_copy['clusters_labels/Predictions']= cluster_labels

# Display sample data
df_scaled_copy.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  clusters_labels/Predictions  
coin_id                                                                
bitcoin                        -0.251637                            3  
ethereum                       -0.199352                            3  
tether                         -0.282061                            1  
ripple                         -0.295546                            1  
bitcoin-cash                   -0.270317                            7

In [59]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
scatter_plot = df_scaled_copy.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    c="cluster_labels/Predictions",  
    cmap="viridis",     
    title="Cryptocurrency Clusters",
    hover_cols=["coin_id"],  
    width=800,
    height=500
)

# Show the scatter plot
scatter_plot


:Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [60]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [61]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
pca_components = pca.fit_transform(df_scaled)

# Create a DataFrame from the transformed components
df_pca = pd.DataFrame(data=pca_components, columns=['PC1', 'PC2', 'PC3'])

# View the first five rows of the DataFrame. 
df_pca.head(5)

PC1       PC2       PC3
0 -0.600667  0.842760  0.461595
1 -0.458261  0.458466  0.952877
2 -0.433070 -0.168126 -0.641752
3 -0.471835 -0.222660 -0.479053
4 -1.157800  2.041209  1.859715

In [62]:
# Retrieve the explained variance to determine how much information can be attributed to each principal component.
data_pca_variance = pca.explained_variance_ratio_

# Display the explained variance ratios for each principal component
for i, explained_variance_ratio in enumerate(data_pca_variance):
    print(f"Explained Variance for PC{i+1}: {explained_variance_ratio:.4f}")
    
# Sum the explained variance ratios for the three principal components
total_explained_variance = sum(data_pca_variance[:3])

# Display the total explained variance
print(f"Total Explained Variance for the Three Principal Components: {total_explained_variance:.4f}")

Explained Variance for PC1: 0.3720
Explained Variance for PC2: 0.3470
Explained Variance for PC3: 0.1760
Total Explained Variance for the Three Principal Components: 0.8950


#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 0.8950

In [63]:
# Create a new DataFrame with the PCA data.
df_market_data_pca = pd.DataFrame(data=pca_components, columns= ['PCA1', 'PCA2', 'PCA3'])


# Copy the crypto names from the original data
df_market_data_pca['con_id'] = df_market_data.index

# Set the coinid column as index
df_market_data_pca.set_index('con_id', inplace=True)

# Display sample data
df_market_data_pca.head()

PCA1      PCA2      PCA3
con_id                                    
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the PCA Data

In [64]:
# Create a list with the number of k-values from 1 to 11
k_values_pca = list(range(1,12))

print(k_values_pca)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]


In [65]:
# Create an empty list to store the inertia values\
inertia_values_pca =[]


# Create a for loop to compute the inertia with each possible value of k
for k in k_values:
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
    kmeans_model = KMeans(n_clusters=k, random_state=1 , n_init=10)
# 2. Fit the model to the data using `df_market_data_pca`
    kmeans_model.fit(df_market_data_pca)
# 3. Append the model.inertia_ to the inertia list
    inertia_values_pca.append(kmeans_model.inertia_)

print(inertia_values_pca)


c:\Users\artur\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\artur\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\artur\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\artur\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Window

[256.87408556789256, 165.90199402036012, 93.77462568057298, 49.66549665179737, 38.352251216386826, 27.61897178795743, 21.134056037473602, 17.437664237020407, 13.742791960480561, 10.48489048597692, 8.114985790821036]


In [66]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_pca = {
    'k_values': k_values_pca,
    'inertia_values': inertia_values_pca
}
# Create a DataFrame with the data to plot the Elbow curve
df_elbow_curve_pca =pd.DataFrame(elbow_data_pca)
df_elbow_curve_pca

k_values  inertia_values
0          1      256.874086
1          2      165.901994
2          3       93.774626
3          4       49.665497
4          5       38.352251
5          6       27.618972
6          7       21.134056
7          8       17.437664
8          9       13.742792
9         10       10.484890
10        11        8.114986

In [67]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
df_elbow_curve_pca.hvplot.line(
    x='k_values',
    y='inertia_values',
    title='Elbow Curve',
    xticks=k_values_pca
)

:Curve   [k_values]   (inertia_values)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer: No 

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [68]:
# Initialize the K-Means model using the best value for k
k_means_model_pca = KMeans(n_clusters=4, random_state=1, n_init=10)

In [69]:
# Fit the K-Means model using the PCA data
k_means_model_pca.fit(df_market_data_pca)

c:\Users\artur\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4, n_init=10, random_state=1)

In [70]:
# Predict the clusters to group the cryptocurrencies using the PCA data
cluster_labels_pca = k_means_model_pca.predict(df_market_data_pca)

# Print the resulting array of cluster values.
cluster_labels_pca


array([1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 1, 0, 0, 3, 0, 0, 0, 0])

In [71]:
# Create a copy of the DataFrame with the PCA data
df_market_data_pca_copy = df_market_data_pca.copy()

# Add a new column to the DataFrame with the predicted clusters
df_market_data_pca_copy['cluster_labels/Predictions'] = cluster_labels_pca

# Display sample data
df_market_data_pca_copy.head()


PCA1      PCA2      PCA3  cluster_labels/Predictions
con_id                                                                
bitcoin      -0.600667  0.842760  0.461595                           1
ethereum     -0.458261  0.458466  0.952877                           1
tether       -0.433070 -0.168126 -0.641752                           0
ripple       -0.471835 -0.222660 -0.479053                           0
bitcoin-cash -1.157800  2.041209  1.859715                           1

In [72]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
scatter_plot_pca = df_market_data_pca_copy.hvplot.scatter(
    x="PCA1",
    y="PCA2",
    c="cluster_labels/Predictions",
    title="Cryptocurrency Clusters (PCA)",
    hover_cols=["coin_id"],
    width=800,
    height=500
)

# Show the scatter plot
scatter_plot_pca

:Scatter   [PCA1]   (PCA2,cluster_labels/Predictions)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [73]:
# Composite plot to contrast the Elbow curves
plot_1 = df_elbow_curve.hvplot.line(x='k_values', y='inertia_values', title='Elbow Curve', xticks=k_values, label = 'Plot 1')
plot_2 = df_elbow_curve_pca.hvplot.line(x='k_values', y='inertia_values', title='Elbow Curve', xticks=k_values_pca, label = 'Plot 2')

composite_plot = plot_1 * plot_2
composite_plot

:Overlay
   .Curve.Plot_1 :Curve   [k_values]   (inertia_values)
   .Curve.Plot_2 :Curve   [k_values]   (inertia_values)

In [78]:
# Composite plot to contrast the clusters
plot_1_cluster_1 = df_scaled_copy.hvplot.scatter(x = "price_change_percentage_24h", y = "price_change_percentage_7d", by = "cluster_labels/Predictions", hover_cols=["coin_id"], marker="o", label = 'Plot 1')
plot_2_cluster_2 = df_market_data_pca_copy.hvplot.scatter(x = "PCA1", y = "PCA2", by = "cluster_labels/Predictions", hover_cols=["coin_id"], marker="x", label = 'Plot 2')

composite_plot_cluster = plot_1_cluster_1 * plot_2_cluster_2
composite_plot_cluster

DataError: Supplied data does not contain specified dimensions, the following dimensions were not found: ['cluster_labels/Predictions']

PandasInterface expects tabular data, for more information on supported datatypes see http://holoviews.org/user_guide/Tabular_Datasets.html

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:There are several factors, for example, by reducing the dimensionality of the data, we can speed up the grouping process, which is that using PCA gives us better grouping, but also, there is the risk of losing valuable information when functions are eliminated, this may result in worse clustering if the removed features were significant.